Add check for local files before processing post images in Photon

hanifn opened the pull request:
#### Summary

In `Jetpack_Photon::filter_the_content`, the call to `strip_image_dimensions_maybe` is outside the `if` clause that checks for local uploads. This is problematic because `strip_image_dimensions_maybe` only checks if the filename has the image dimensions string, e.g.: `300x300`, and then calls `file_exists` on this file after removing the image dimensions and prepending the `basedir`. If a file is hosted externally, like on a CDN, the call to `file_exists` will only return false. Additionally, if the site is using a custom Stream wrapper implementation to upload files to the CDN, these `file_exists` calls results in a lot of requests to the CDN that returns `404` status. This means those requests are uncached and would slow down the site load times.

#### Changes proposed in this Pull Request:
* Create a dedicated function to check if a file is locally uploaded or not.
* Add the local file check into `strip_image_dimensions_maybe`.

#### Testing instructions:
_Note: Tested on a site hosted on VIP Go._

* Enable Jetpack Photon for site.
* Create a new post with an image that's hosted externally and with a filename that has ends in something like `-300x300.jpg`.
* Visit post and make sure there's no external calls to CDN that results in a `404` response.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add check for local file upload before processing post images with Photon

hanifn committed:
Add dedicated function to check for local uploads

hanifn committed:
Add local uploads check when stripping image dimensions

jeherve commented on class.photon.php:
```diff
@@ -358,7 +358,7 @@ public static function filter_the_content( $content ) {
 					// WP Attachment ID, if uploaded to this site
 					if (
 						preg_match( '#class=["|\']?[^"\']*wp-image-([\d]+)[^"\']*["|\']?#i', $images['img_tag'][ $index ], $attachment_id ) &&
-						0 === strpos( $src, $upload_dir['baseurl'] ) &&
```
Since we don't use it anymore in that function, maybe we could remove `$upload_dir = wp_get_upload_dir();` from `filter_the_content`?

hanifn replied:
Nice catch! I've removed the call to `wp_get_upload_dir()`

jeherve commented on class.photon.php:
```diff
@@ -1264,4 +1266,17 @@ public function cleanup_rest_photon_image_downsize( $response ) {
 	}
+
+	/**
+	 * Checks if image is a locally upload file via it's URL
```
```suggestion
	 * Checks if image is a locally uploaded file via its URL
```

jeherve reviewed (dismissed):
It seems to work in my tests. Thank you for adding tests as well. I would have 2 minor remarks.

hanifn committed:
Fix comments in Photon class

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

hanifn committed:
Remove unused call to wp_get_upload_dir

zinigor commented on class.photon.php:
```diff
@@ -1076,7 +1075,9 @@ public static function strip_image_dimensions_maybe( $src ){
-		if ( preg_match( '#(-\d+x\d+)\.(' . implode('|', self::$extensions ) . '){1}$#i', $src, $src_parts ) ) {
+		if ( self::is_local_upload( $src ) &&
+			preg_match( '#(-\d+x\d+)\.(' . implode('|', self::$extensions ) . '){1}$#i', $src, $src_parts ) )
+		{
```
The `if` condition indentation looks weird to me, may I suggest this instead?
```
		if (
			self::is_local_upload( $src ) &&
			preg_match( '#(-\d+x\d+)\.(' . implode('|', self::$extensions ) . '){1}$#i', $src, $src_parts ) 
		) {
```

zinigor reviewed (changes requested):
I know I'm probably overthinking it, but can we go back to only getting upload dir details once per image? Currently with this code we're doing it three times. And while the code only runs the "lightweight" version that doesn't refresh the upload folder every time, it still [runs through filters](https://core.trac.wordpress.org/browser/tags/5.2/src/wp-includes/functions.php#L2095) that may be set on the upload dir location. We will be triggering these filters three times more now with this change.
Can we get rid of the `is_local_upload` method, store the upload folder data as before and just run `strpos` inside if conditions?

hanifn commented:
@zinigor That makes sense. Considering that we're only using this in only 2 places in the class right now, the new `is_local_upload` might be unneeded. Still, we would need to do the same check in `strip_image_dimensions_maybe` so the number of calls to `wp_get_upload_dir()` would only be reduced by one. What do you think of moving the call to `strip_image_dimensions_maybe()` inside the [first `if` clause](https://github.com/Automattic/jetpack/blob/ebe6dc3793bdad659184d6d5868822686b11e355/class.photon.php#L359-L414) that checks for local file? That should reduce the need to get the upload directories to just once.

zinigor commented:
Whoa, sorry for such a delay with the response, @hanifn . Yeah, sure, if we can reduce the number of calls even more, I'm all for that. Can you please make it, as well as rebase the code so that it wouldn't conflict? I'm happy to take a look and test again.

hanifn committed:
Remove is_local_upload function

using is_local_upload results in an extra call to wp_get_upload_dir() every image.

hanifn commented:
@zinigor Changed made and rebased on `master` in a84daa00a22b59cc8d278c5c6441e937db92ef3c

zinigor reviewed (approved):
Thanks, sorry it took me so long to review this one!
